src/VisualStudio/Core/Def/Telemetry/AggregatingTelemetryLogManager.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Telemetry;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Telemetry
+{
+    /// <summary>
+    /// Manages creation and obtaining aggregated telemetry logs. Also, notifies logs to
+    /// send aggregated events every 30 minutes.
+    /// </summary>
+    internal sealed class AggregatingTelemetryLogManager
+    {
+        private static readonly TimeSpan s_batchedTelemetryCollectionPeriod = TimeSpan.FromMinutes(30);
+
+        private readonly TelemetrySession _session;
+        private readonly AsyncBatchingWorkQueue _postTelemetryQueue;
+
+        /// <summary>
+        /// Indicates whether <see cref="_postTelemetryQueue"/> has work already added to it.
+        /// </summary>
+        private int _workAdded = 0;
```

Comment:
not sure why this is needed.  you can always just add more work to the queue.  you're using the no-type-arg version, so it will be a no-op if htere's already an item in the queue.